packages/core/src/services/commits/runDocumentAtCommit.ts

```diff
@@ -5,57 +5,92 @@ import {
+  Workspace,
+} from '@latitude-data/core/browser'
+import { NotFoundError, Result, UnprocessableEntityError } from '$core/lib'
 import { streamToGenerator } from '$core/lib/streamToGenerator'
 import { ProviderApiKeysRepository } from '$core/repositories'
+import { v4 as uuid } from 'uuid'
+import { ZodError } from 'zod'
 
 import { ai, AILog, validateConfig } from '../ai'
 import { getResolvedContent } from '../documents/getResolvedContent'
 
+type CachedApiKeys = Map<string, ProviderApiKey>
+async function cacheApiKeysByName({ workspaceId }: { workspaceId: number }) {
+  const scope = new ProviderApiKeysRepository(workspaceId)
+  const result = await scope.findAll().then((r) => r.unwrap())
+
+  return result.reduce((acc, apiKey) => {
+    acc.set(apiKey.name, apiKey)
+    return acc
+  }, new Map<string, ProviderApiKey>())
+}
+
 export async function runDocumentAtCommit({
-  documentUuid,
+  workspaceId,
+  document,
   commit,
   parameters,
-  logHandler,
+  providerLogHandler,
+  generateUUID = uuid,
 }: {
-  documentUuid: string
+  workspaceId: Workspace['id']
+  document: DocumentVersion
   commit: Commit
   parameters: Record<string, unknown>
-  logHandler: (log: AILog) => void
+  providerLogHandler: (log: AILog) => void
+  generateUUID?: typeof uuid
 }) {
-  const workspace = await findWorkspaceFromCommit(commit)
-  if (!workspace) throw Result.error(new NotFoundError('Workspace not found'))
-
   const result = await getResolvedContent({
-    workspace,
-    documentUuid,
+    workspaceId,
+    document,
     commit,
   })
 
   if (result.error) return result
 
-  const chain = createChain({ prompt: result.value, parameters })
-  const scope = new ProviderApiKeysRepository(workspace.id)
+  const resolvedContent = result.value
+  const chain = createChain({ prompt: resolvedContent, parameters })
+  const documentLogUuid = generateUUID()
 
-  let stream: ReadableStream
-  let response: Promise<ChainCallResponse> | undefined = undefined
+  let responseResolve: (value: ChainCallResponse) => void
+  let responseReject: (reason?: any) => void
 
-  await new Promise<void>((resolve) => {
-    stream = new ReadableStream<ChainEvent>({
-      start(controller) {
-        response = iterate({ chain, scope, controller, logHandler })
+  const response = new Promise<ChainCallResponse>((resolve, reject) => {
+    responseResolve = resolve
+    responseReject = reject
+  })
 
-        resolve()
-      },
-    })
+  const allApiKeys = await cacheApiKeysByName({ workspaceId })
+  const stream = new ReadableStream<ChainEvent>({
+    start(controller) {
+      iterate({
+        chain,
+        allApiKeys,
+        controller,
+        providerLogHandler,
+        documentLogUuid,
+      })
+        .then(responseResolve)
+        .catch(responseReject)
+    },
   })
 
+  // Dummy handling of the response
+  // This is helpful for not throwing the error
+  // when no one is listening to the promise
+  response.then(() => {}).catch(() => {})
```

Comment:
Response is ready when stream finish. We do this here because if an error is thrown and the consumer is not using the response like we are doing in API gateway the error is raised and appears on API logs. This way it does not appear and also consumer can `catch` the error if using the `response`.